### PR TITLE
Improve R Tracking API

### PR DESCRIPTION
**Pull request by kevinykuo**
This PR exposes the previously unexported tracking client API and attains parity with the REST and Python interfaces.

The "fluent" and "client" features are unified into one set of functions. When provided, the client object is be respected (similar to Python's `mlflow.tracking`). For functions concerned with runs, an active run is started and registered if `run_id` is not specified (similar to the fluent API).

Data structures returned by tracking functions are standardized to be tidy tibbles. Each experiment is a row, each run is a row, and each data element record (metric/param/tag) is a row. This enables easier inspection and further manipulation compared to nested lists.

## Examples

We can create an experiment by just passing the name, and the client is inferred (respects `mlflow_set_tracking_uri()` and the `MLFLOW_TRACKING_URI` environment variable.)
```
experiment <- mlflow_create_experiment("myexp")
experiment

# A tibble: 1 x 4
  experiment_id name  artifact_location                     lifecycle_stage
  <chr>         <chr> <chr>                                 <chr>          
1 1             myexp /Users/kevinykuo/Dropbox/Projects/ml… active    
```

Similarly, we can create a run by providing the experiment ID.

```
experiment_id <- experiment$experiment_id
run <- mlflow_create_run(experiment$experiment_id)
dplyr::glimpse(run)

Observations: 1
Variables: 13
$ run_uuid        <chr> "a15b485c7fd0482c8b6cce302c581196"
$ experiment_id   <chr> "1"
$ name            <chr> ""
$ source_type     <chr> "NOTEBOOK"
$ source_name     <chr> ""
$ user_id         <chr> "kevinykuo"
$ status          <chr> "RUNNING"
$ start_time      <dttm> 2019-04-12 02:37:35
$ artifact_uri    <chr> "/Users/kevinykuo/Dropbox/Projects/mlflow-ex/mlruns/1/a15b485c7fd0482c8…
$ lifecycle_stage <chr> "active"
$ metrics         <lgl> NA
$ params          <lgl> NA
$ tags            <lgl> NA
```

Log some metrics and inspect.

```
run_id <- run$run_uuid
mlflow_log_metric("mse", 24, run_id = run_id)
mlflow_log_metric("mse", 25, run_id = run_id)

mlflow_get_metric_history(run_id, "mse")

# A tibble: 2 x 3
  key   value timestamp          
  <chr> <dbl> <dttm>             
1 mse      24 2019-04-12 02:38:47
2 mse      25 2019-04-12 02:38:47
```

We can also explicitly pass a client.

```
client <- mlflow_client()

mlflow_log_metric("mse", 26, client = client, run_id = run_id)
mlflow_get_metric_history(client = client, run_id, "mse")

# A tibble: 3 x 3
  key   value timestamp          
  <chr> <dbl> <dttm>             
1 mse      24 2019-04-12 02:38:47
2 mse      25 2019-04-12 02:38:47
3 mse      26 2019-04-12 02:39:21
```

When only `client` is provided, we require a `run_id` to avoid ambiguity.

```
mlflow_log_metric("mse", 27, client = client)
## Error: `run_id` must be specified when `client` is specified. 
```

Each `RunInfo` is represented as a row in the tibble.

```
mlflow_list_run_infos(experiment$experiment_id)
# A tibble: 1 x 10
  run_uuid experiment_id name  source_type source_name user_id status start_time         
  <chr>    <chr>         <chr> <chr>       <chr>       <chr>   <chr>  <dttm>             
1 a15b485… 1             ""    NOTEBOOK    ""          keviny… RUNNI… 2019-04-12 02:37:35
# … with 2 more variables: artifact_uri <chr>, lifecycle_stage <chr>
```

For batch logging, we take named lists, which is more natural than requiring nested lists with `key` and `value`. An optional param `timestamps` can be used if metrics need to be associated with times that are not the current time.
```
run <- mlflow_create_run(experiment$experiment_id)
mlflow_log_batch(
  metrics = list(rmse = 24, auc = 0.99), 
  params = list(alpha = 0.5, lambda = 0.5),
  tags = list(foo = "bar", bar = "baz"),
  run_id = run$run_uuid
)
```

Again, each run is a row:

```
run <- mlflow_get_run(run$run_uuid)

> dplyr::glimpse(run)
Observations: 1
Variables: 13
$ run_uuid        <chr> "0abd954521a94bf095569250a669dfcb"
$ experiment_id   <chr> "1"
$ name            <chr> ""
$ source_type     <chr> "NOTEBOOK"
$ source_name     <chr> ""
$ user_id         <chr> "kevinykuo"
$ status          <chr> "RUNNING"
$ start_time      <dttm> 2019-04-12 02:40:18
$ artifact_uri    <chr> "/Users/kevinykuo/Dropbox/Projects/mlflow-ex/mlruns/1/0abd954521a94bf09…
$ lifecycle_stage <chr> "active"
$ metrics         <list> [<tbl_df[2 x 3]>]
$ params          <list> [<tbl_df[2 x 2]>]
$ tags            <list> [<tbl_df[2 x 2]>]


run$metrics

[[1]]
# A tibble: 2 x 3
  key   value timestamp          
  <chr> <dbl> <dttm>             
1 auc    0.99 2019-04-12 02:40:18
2 rmse  24    2019-04-12 02:40:18
```

### CODE REVIEW

**Review by tomasatdatabricks — COMMENTED** [review]
Looks good, thanks Kevin!

a) I have a couple of high level questions:
1.  mlflow_start_run vs mlflow_create_run - can / should we just keep one of them? mlflow_start_run?
2. Do we want to use active experiment / run id to fill in the arguments whenever possible?
    - I think this would be ok and simplify the behavior except for delete / restore calls. I wonder if we can just treat these as special case and throw exception if you are attempting to delete active run / experiment?
3. [minor] For all the apis that use mlflow_client = NULL, should we delay the creation of the client until it's actually needed? I.e. just pass the client or NULL along.

b) It loos like there is a broken merge in tracking-rest.R, I left a comment in code.

File: mlflow/R/mlflow/R/tracking-methods.R [hunk 1]
```diff
@@ -0,0 +1,602 @@
...
+#'   During a single execution of a run, a particular metric can be logged several times.
+#'   Backend will keep track of historical values along with timestamps.
+#'
+#' @param key Name of the metric.
+#' @param value Float value for the metric being logged.
+#' @param timestamp Unix timestamp in milliseconds at the time metric was logged.
+#' @template roxlate-run-id
+#' @template roxlate-client
+#' @export
+mlflow_log_metric <- function(key, value, timestamp = NULL, client = NULL, run_id = NULL) {
```
[the remark this answers is not in the document]
Comment (kevinykuo) [reply mlflow/R/mlflow/R/tracking-methods.R]: Done!

File: mlflow/R/mlflow/R/tracking-rest.R [hunk 2]
```diff
@@ -53,6 +54,10 @@ mlflow_rest <- function( ..., client, query = NULL, data = NULL, verb = "GET", v
     mlflow_rest_path(version),
     paste(args, collapse = "/")
   )
+
+  config <- config(ssl_verifypeer = rest_config$verify_peer)
```
[the remark this answers is not in the document]
Comment (kevinykuo) [reply mlflow/R/mlflow/R/tracking-rest.R]: Failed git-fu on my part, should be fixed.

File: mlflow/R/mlflow/R/tracking-utils.R [hunk 3]
```diff
@@ -101,3 +101,41 @@ MLFLOW_SOURCE_TYPE <- list(
...
+resolve_client_and_run_id <- function(client, run_id) {
+  if (is.null(client)) {
+    if (is.null(run_id)) {
+      run_id <- run_id(mlflow_get_or_start_run())
+    }
+    client <- mlflow_client()
+  } else {
+    if (is.null(run_id)) stop("`run_id` must be specified when `client` is specified.", call. = FALSE)
+  }
+  list(client, run_id)
```
[the remark this answers is not in the document]
Comment (kevinykuo) [reply mlflow/R/mlflow/R/tracking-utils.R]: OK

File: mlflow/R/mlflow/R/tracking-client.R [hunk 4]
```diff
@@ -114,381 +117,10 @@ new_mlflow_client.mlflow_https <- function(tracking_uri) {
 #'
 #' @param tracking_uri The tracking URI. If not provided, defaults to the service
 #'  set by `mlflow_set_tracking_uri()`.
-#' @keywords internal
+#' @export
 mlflow_client <- function(tracking_uri = NULL) {
   tracking_uri <- new_mlflow_uri(tracking_uri %||% mlflow_get_tracking_uri())
   client <- new_mlflow_client(tracking_uri)
   mlflow_validate_server(client)
```
[the remark this answers is not in the document]
Comment (kevinykuo) [reply mlflow/R/mlflow/R/tracking-client.R]: Updated to validate only for file client

**Comment by kevinykuo** [comment]
@tomasatdatabricks Thanks! Comments have been addressed and some thoughts re: your questions below.
- On `mlflow_create_run()` and `mlflow_start_run()` -- I think we should keep both. Users are used to the latter starting an *active* run. Also the Python docs has been [explicit in saying that "create run" does not register the active run](https://mlflow.org/docs/latest/python_api/mlflow.tracking.html#mlflow.tracking.MlflowClient.create_run).
- We'll fill in active experiment/run IDs whenever possible. Informative errors are given when user tries to delete an active experiment.
- The client is always needed because we always hit the REST API.

**Comment by tomasatdatabricks** [comment]
Hey Kevin, thanks for the updates addressing my comments!

I have some more questions regarding mlflow_create_run / mlflow_start_run. 

a) Personally I think it's a  bit weird that mfllow_start_run infers some information from the context by calling mlflow_get_run_context, whereas create run does not. I.e. the difference between the two is not just whether the run is started / activated. I don't think we should be calling get_run_context in create run since create run is supposed to give user full control over what is being recorded. The same is true in python. However, python client has a clear distinction between client and fluent apis and therefore it's a bit more obvious that these are different functions. Is there a good way how we can make this distinction more obvious in R? Maybe we can refactor the code so that get_run_context is called inside mlflow_create_run but user can override it by passing in context.

b) Can we somehow make it clear that you do not have call mlflow_create_run in order to start one? I am worried people might understand it that way.

I would consider grouping the more lower level apis together somehow. E.g. with mlflow_manage prefix. I think it would be easier to explain to people that mlflow_manage is not meant to be used on regular basis.

What do you think?

**Comment by kevinykuo** [comment]
Removed both `mlflow_create_run()` and `mlflow_set_terminated()` in favor of `mlflow_start/end_run()`.

File: mlflow/R/mlflow/R/tracking-runs.R [hunk 5]
```diff
@@ -0,0 +1,597 @@
...
+#' @export
+mlflow_end_run <- function(status = c("FINISHED", "SCHEDULED", "FAILED", "KILLED"),
+                           end_time = NULL, run_id = NULL, client = NULL) {
+
+  status <- match.arg(status)
+  end_time <- end_time %||% current_time()
+
+  active_run_id <- mlflow_get_active_run_id()
+
+  if (!is.null(client) && is.null(run_id))
```
Comment (tomasatdatabricks) [inline mlflow/R/mlflow/R/tracking-runs.R]: should we just call resolve_client_and_run_id?

File: mlflow/R/mlflow/R/tracking-runs.R [hunk 6]
```diff
@@ -0,0 +1,597 @@
...
+#' @param source_type Integer enum value describing the type of the run  ("local", "project", etc.).
+#' @param user_id User ID or LDAP for the user executing the run. Only used when `client` is specified.
+#' @param run_name Human readable name for run. Only used when `client` is specified.
+#' @param start_time Unix timestamp of when the run started in milliseconds. Only used when `client` is specified.
+#' @param tags Additional metadata for run in key-value pairs. Only used when `client` is specified.
+#' @template roxlate-client
+#'
+#' @examples
+#' \dontrun{
+#' with(mlflow_start_run(), {
```
Comment (tomasatdatabricks) [inline mlflow/R/mlflow/R/tracking-runs.R]: should this be with.mlflow_run?

File: mlflow/R/mlflow/R/tracking-experiments.R [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+#' Gets metadata for an experiment and a list of runs for the experiment.
+#'
+#'
+#' @param name The experiment name, either this or `experiment_id` should be specified.
+#' @param experiment_id Identifer to get an experiment. Attempts to obtain the active experiment
+#'   if not provided.
+#' @template roxlate-client
+#' @export
+mlflow_get_experiment <- function(name = NULL, experiment_id = NULL, client = NULL) {
+  if (!is.null(name) && !is.null(experiment_id)) {
```
Comment (tomasatdatabricks) [inline mlflow/R/mlflow/R/tracking-experiments.R]: should we attempt to infer the experiment if no id nor name is provided?

File: mlflow/R/mlflow/R/tracking-client.R [hunk 8]
```diff
@@ -4,6 +4,9 @@ new_mlflow_client <- function(tracking_uri) {
 }
 
 new_mlflow_uri <- function(raw_uri) {
+  if (!grepl("://", raw_uri)) {
+    raw_uri <- paste0("file://", raw_uri)
```
Comment (tomasatdatabricks) [inline mlflow/R/mlflow/R/tracking-client.R]: we need special case for "databricks".

File: mlflow/R/mlflow/tests/testthat/test-databricks-utils.R [hunk 9]
```diff
@@ -14,7 +14,7 @@ test_that("mlflow creates databricks client when scheme is databricks", {
       expect_true(paste(args, collapse = "/") == "experiments/list")
       list(experiments = c(1, 2, 3))
     }, {
-      client1 <- mlflow:::mlflow_client("databricks")
+      client1 <- mlflow:::mlflow_client("databricks://")
```
Comment (tomasatdatabricks) [inline mlflow/R/mlflow/tests/testthat/test-databricks-utils.R]: This needs to remain "databricks".

**Review by tomasatdatabricks — CHANGES_REQUESTED** [review]
The PR looks really good, thanks Kevin!

I had a few small comments / questions in code  + the following:

1. We need a special case for tracking_uri == "databricks". 
2. I did not see any tests for mlflow_log_batch? We need to add some or take the log_batch out and make a follow up PR.

I still want to double check test coverage for the rest.

File: mlflow/R/mlflow/tests/testthat/test-databricks-utils.R (hunk 9 above)
Comment (kevinykuo) [reply mlflow/R/mlflow/tests/testthat/test-databricks-utils.R]: Will add special case for this.

File: mlflow/R/mlflow/R/tracking-experiments.R [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+  response <- mlflow_rest(
+    "experiments", "get",
+    client = client, query = list(experiment_id = experiment_id)
+  )
+  response$experiment %>%
+    new_mlflow_experiment()
+}
+
+mlflow_get_experiment_by_name <- function(name, client = NULL) {
+  client <- client %||% mlflow_client()
```
[the remark this answers is not in the document]
Comment (kevinykuo) [reply mlflow/R/mlflow/R/tracking-experiments.R]: Ideally we should check all user inputs.

File: mlflow/R/mlflow/R/tracking-experiments.R [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+#' @export
+mlflow_set_experiment <- function(experiment_name = NULL, experiment_id = NULL, artifact_location = NULL) {
+  if (!is.null(experiment_name) && !is.null(experiment_id)) {
+    stop("Only one of `experiment_name` or `experiment_id` should be specified.",
+         call. = FALSE
+    )
+  }
+
+  if (is.null(experiment_name) && is.null(experiment_id)) {
+    stop("At least one of `experiment_name` or `experiment_id` should be specified.",
```
[the remark this answers is not in the document]
Comment (kevinykuo) [reply mlflow/R/mlflow/R/tracking-experiments.R]: Will fix.

File: mlflow/R/mlflow/R/tracking-experiments.R (hunk 7 above)
Comment (kevinykuo) [reply mlflow/R/mlflow/R/tracking-experiments.R]: yeah I think it makes sense to return the active experiment or whatever the user set in the env variable?

File: mlflow/R/mlflow/R/tracking-runs.R (hunk 6 above)
Comment (kevinykuo) [reply mlflow/R/mlflow/R/tracking-runs.R]: `mlflow_start_run()` returns an `mlflow_run` object so `with` will dispatch accordingly

File: mlflow/R/mlflow/R/tracking-runs.R (hunk 5 above)
Comment (kevinykuo) [reply mlflow/R/mlflow/R/tracking-runs.R]: `resolve_client_and_run_id()` could start a run though

File: mlflow/R/mlflow/R/tracking-experiments.R [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+  if (identical(experiment_id, mlflow_get_active_experiment_id()))
+    stop("Cannot delete an active experiment.", call. = FALSE)
+
+  client <- client %||% mlflow_client()
+  mlflow_rest(
+    "experiments", "delete",
+    verb = "POST", client = client,
+    data = list(experiment_id = experiment_id)
+  )
+  mlflow_get_experiment(experiment_id = experiment_id)
```
[the remark this answers is not in the document]
Comment (kevinykuo) [reply mlflow/R/mlflow/R/tracking-experiments.R]: We're generally following the rule that operations on experiments return experiments, operations on runs return runs, (and logging operations are currently inconsistent.) I think it makes sense to give the user feedback right away on the side effect, and if we're printing it we might as well return it.

**Comment by kevinykuo** [comment]
Will add a test for `mlflow_log_batch()`.

File: mlflow/R/mlflow/R/tracking-experiments.R [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+  )
+  experiment <- mlflow_get_experiment(experiment_id)
+
+  experiment
+}
+
+#' Set Experiment
+#'
+#' Sets an experiment as the active experiment. If the experiment does not exist,
+#'   creates an experiment with provided name.
```
[the remark this answers is not in the document]
Comment (kevinykuo) [reply mlflow/R/mlflow/R/tracking-experiments.R]: Will fix.

File: mlflow/R/mlflow/R/tracking-experiments.R [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+  mlflow_set_active_experiment_id(mlflow_id(experiment))
+  experiment
+}
+
+#' Active Experiment
+#'
+#' Retrieves the active experiment.
+#'
+#' @export
+mlflow_active_experiment <- function() {
```
[the remark this answers is not in the document]
Comment (kevinykuo) [reply mlflow/R/mlflow/R/tracking-experiments.R]: If we have `mlflow_active_run()` it's reasonable to assume we have `mlflow_active_experiment()`? One question here is whether a provided `MLFLOW_EXPERIMENT_NAME/ID` should be returned if an experiment hasn't been "set" (via `mlflow_set_experiment()`).

File: mlflow/R/mlflow/R/tracking-globals.R [hunk 15]
```diff
@@ -1,27 +1,19 @@
-#' Active Run
-#'
-#' Retrieves the active run.
-#'
-#' @export
-mlflow_active_run <- function() {
-  .globals$active_run
-}
+mlflow_get_run_id <- function(run) cast_nullable_string(run$run_uuid)
```
[the remark this answers is not in the document]
Comment (kevinykuo) [reply mlflow/R/mlflow/R/tracking-globals.R]: This function isn't used, will remove.

File: mlflow/R/mlflow/R/tracking-runs.R [hunk 16]
```diff
@@ -0,0 +1,597 @@
...
+    key = key,
+    value = value,
+    timestamp = timestamp
+  ))
+  invisible(value)
+}
+
+
+
+
```
[the remark this answers is not in the document]
Comment (kevinykuo) [reply mlflow/R/mlflow/R/tracking-runs.R]: OK

**Comment by kevinykuo** [comment]
@tomasatdatabricks 
- Removed `mlflow_active_run()` and `mlflow_active_experiment()` in favor of `mlflow_get_run()` and `mlflow_get_experiment()`. 
- Special case `databricks` URI so `://` isn't needed.
- Added tests for `mlflow_log_batch()`.
- Added suggested docs/error message improvements.

**Comment by kevinykuo** [comment]
@tomasatdatabricks can we merge?

File: mlflow/R/mlflow/R/tracking-experiments.R (hunk 7 above)
Comment (tomasatdatabricks) [reply mlflow/R/mlflow/R/tracking-experiments.R]: Yeah, sounds good to me.

File: mlflow/R/mlflow/R/tracking-experiments.R (hunk 12 above)
[the remark this answers is not in the document]
Comment (tomasatdatabricks) [reply mlflow/R/mlflow/R/tracking-experiments.R]: Ok, sounds good for now but I think we should revisit this before 1.0. There can be some subtle differences in behavior compared to python.

File: mlflow/R/mlflow/R/tracking-globals.R (hunk 15 above)
[the remark this answers is not in the document]
Comment (tomasatdatabricks) [reply mlflow/R/mlflow/R/tracking-globals.R]: @kevinykuo can you remove it?

**Review by tomasatdatabricks — CHANGES_REQUESTED** [review]
There is one last nit that needs to be resolved, after that we can merge.

**Comment by kevinykuo** [comment]
Oops missed that one, fixed!

**Review by tomasatdatabricks — APPROVED** [review]
lgtm!

**Pull request merged**